h2o-core/src/main/java/water/DKVManager.java

```diff
@@ -0,0 +1,122 @@
+   *
+   * @param retainedKeys Keys of {@link Frame}s and {@link Model}s to be retained.
+   */
+  public static void retain(final Key[] retainedKeys){
+    final Set<Key> retainedSet = new HashSet<>(retainedKeys.length);
+    retainedSet.addAll(Arrays.asList(retainedKeys));
+    extractNestedKeys(retainedSet);
+
+    new ClearDKVTask(retainedSet.toArray(new Key[]{}))
+            .doAllNodes();
+  }
+
+  /**
+   * Iterates through the keys provided by the user, dropping any keys that are not a Model key or a Frame key.
+   * Afterwards, extracts
+   *
+   * @param retainedKeys A {@link Set} of retained keys to insert the extracted {@link Frame} and {@link Model} keys to.
+   *                     Should contain user-specified keys to retain in order to extract anything.
+   */
+  private static void extractNestedKeys(final Set<Key> retainedKeys) {
+
+    final Iterator<Key> keysIterator = retainedKeys.iterator(); // Traverse keys provided by the user only.
+    while (keysIterator.hasNext()) {
+      final Key key = keysIterator.next();
+
+      final Value value = DKV.get(key);
+
+      if (value == null || value.isNull()) {
+        continue; // The key might not have value cached locally. Do not delete.
+      } else if (!value.isFrame() && !value.isModel()) {
+        Log.info(String.format("Given key %s is of type %d. Not retaining.", key.toString(), value.type()));
+      } else if (value.isFrame()) {
+        extractFrameKeys(retainedKeys, (Frame) value.get());
+      } else if (value.isModel()) {
+        extractModelKeys(retainedKeys, (Model) value.get());
+      }
+    }
+  }
+
+  /**
+   * Extracts keys a {@link Frame} points to.
+   *
+   * @param retainedkeys A set of retained keys to insert the extracted {@link Frame} keys to.
+   * @param frame        An instance of {@link Frame} to extract the keys from.
+   */
+  private static void extractFrameKeys(final Set<Key> retainedkeys, final Frame frame) {
+    Objects.requireNonNull(frame);
+    final Key<Vec>[] frameKeys = frame.keys();
+    for (Key k : frameKeys) {
+      retainedkeys.add(k);
+    }
+  }
+
+  /**
+   * Exctracts keys a {@link Model} points to.
+   *
+   * @param retainedKeys A set of retained keys to insert the extracted {@link Model} keys to.
+   * @param model        An instance of {@link Model} to extract the keys from
+   */
+  private static void extractModelKeys(final Set<Key> retainedKeys, final Model model) {
+    Objects.requireNonNull(model);
+    if (model._parms._train != null) {
+      retainedKeys.add(model._parms._train);
+      extractFrameKeys(retainedKeys, model._parms._train.get());
+    }
+    if (model._parms._valid != null) {
+      retainedKeys.add(model._parms._valid);
+      extractFrameKeys(retainedKeys, model._parms._valid.get());
+    }
+  }
+
+  private static class ClearDKVTask extends MRTask<ClearDKVTask> {
+
+    private final Key[] _retainedKeys; // Only model and frame keys
```

Comment:
is this true? how about the shared vecs?

Reply:
No longer true, old commentary. Updated the commentary. Resolved in https://github.com/h2oai/h2o-3/pull/3577/commits/3786ef7de44574103c72247aba21c48e07126e67